tool/build/renamestr.c

```diff
@@ -0,0 +1,469 @@
+static wontreturn void DieSys(const char *thing) {
+  perror(thing);
+  exit(1);
+}
+
+static wontreturn void ShowUsage(int rc, int fd) {
+  tinyprint(fd, "USAGE\n\n  ", prog, MANUAL, NULL);
+  exit(rc);
+}
+
+static wontreturn void DieOom(void) {
+  Die("renamestr", "out of memory");
+}
+
+static void *Malloc(size_t n) {
+  void *p;
+  if (!(p = malloc(n)))
+    DieOom();
+  return p;
+}
+
+static void *Realloc(void *p, size_t n) {
+  if (!(p = realloc(p, n)))
+    DieOom();
+  return p;
+}
+
+static void Pwrite(const void *data, size_t size, uint64_t offset) {
+  ssize_t rc;
+  const char *p, *e;
+  for (p = data, e = p + size; p < e; p += (size_t)rc, offset += (size_t)rc) {
+    if ((rc = pwrite(exefd, p, e - p, offset)) == -1) {
+      DieSys(exepath);
+    }
+  }
+}
+
+struct Param {
+  const char *from_string;
+  size_t from_len;
+  const char *to_string;
+  size_t to_len;
+  int count;
+  char *roloc;
+};
+
+struct Params {
+  int n;
+  struct Param p[4];
+};
+
+static struct Params params;
+
+static void GetOpts(int argc, char *argv[]) {
+  int opt;
+  bool partial = false;
+  params.n = 0;
+  struct Param *param;
+  while ((opt = getopt(argc, argv, "hvf:t:")) != -1) {
+    if (params.n >= ARRAYLEN(params.p)) {
+      param = NULL;
+    } else {
+      param = &(params.p[params.n]);
+    }
+    switch (opt) {
+      case 'f':
+        if (!param) {
+          Die(prog, "too many replacements provided");
+        }
+        if (param->from_string) {
+          Die(prog, "from string already provided");
+        }
+        param->from_string = optarg;
+        param->from_len = strlen(optarg);
+        partial = !partial;
+        break;
+      case 't':
+        if (!param) {
+          Die(prog, "too many replacements provided");
+        }
+        if (param->to_string) {
+          Die(prog, "to string already provided");
+        }
+        param->to_string = optarg;
+        param->to_len = strlen(optarg);
+        partial = !partial;
+        break;
+      case 'v':
+        tinyprint(0, VERSION, NULL);
+        exit(0);
+      case 'h':
+        ShowUsage(0, 1);
+      default:
+        ShowUsage(1, 2);
+    }
+    if (param->from_string && param->to_string) {
+      if (param->from_len < param->to_len) {
+        Die(prog, "to_string longer than from_string, cannot replace");
+      }
+      params.n++;
+    }
+  }
+  if (params.n == 0) {
+    Die(prog, "no replacements provided");
+  }
+  if (partial) {
+    Die(prog, "partial replacement provided");
+  }
+  for (int i = 0; i < params.n; ++i) {  // no-op
+    if (!params.p[i].from_string) {
+      Die(prog, "need from_string for replacement");
+    }
+    if (!params.p[i].to_string) {
+      Die(prog, "need to_string for replacement");
+    }
+  }
+  if (optind == argc) {
+    Die(prog, "missing input argument");
+  }
+  if (optind != argc - 1) {
+    Die(prog, "too many args");
+  }
+  exepath = argv[optind];
+}
+
+static void ValidateElfImage(Elf64_Ehdr *e, Elf64_Off esize,  //
```

Comment:
Don't bother validating. Your program doesn't even use program headers. You only care about sections. Some of these restrictions, like that it has to be a non-dynamic executable, don't make sense. Since your program could generalize to object files too.